ffi/src/lib.rs

```diff
@@ -620,30 +620,136 @@ pub unsafe extern "C" fn version(snapshot: *const SnapshotHandle) -> u64 {
     snapshot.version()
 }
 
+/// The `EngineSchemaVisitor` defines a visitor system to allow engines to build their own
+/// representation of a schema from a particular schema within kernel.
+///
+/// The model is list based. When the kernel needs a list, it will ask engine to allocate one of a
+/// particular size. Once allocated the engine returns an `id`, which can be any identifier the
+/// engine wants, and will be passed back to the engine to identify the list in the future.
+///
+/// Every schema element the kernel visits belongs to some list of "sibling" elements. The schema
+/// itself is a list of schema elements, and every complex type (struct, map, array) contains a list
+/// of "child" elements.
+///  1. Before visiting schema or any complex type, the kernel asks the engine to allocate a list to
+///     hold its children
+///  2. When visiting any schema element, the kernel passes its parent's "child list" as the
+///     "sibling list" the element should be appended to:
+///      - For the top-level schema, visit each top-level column, passing the column's name and type
+///      - For a struct, first visit each struct field, passing the field's name, type, nullability,
+///        and metadata
+///      - For a map, visit the key and value, passing its special name ("map_key" or "map_value"),
+///        type, and value nullability (keys are never nullable)
+///      - For a list, visit the element, passing its special name ("array_element"), type, and
+///        nullability
```

Comment:
nit: not sure if it makes that much of a difference, but should be we consistent with the arrow conversion in how we name the special fields?

Reply:
I looked about and couldn't find any consistency online about what to name these. Agree that it's not great to have different names here than in our `schema.rs` code.

In the spirit of not bikeshedding too much I've opened #202 and will follow-up to rename everything properly.